networkx/lazy_imports.py

```diff
@@ -0,0 +1,179 @@
+        # e.g. ``module.__spec__.loader = None`` would trigger a load from
+        # trying to access module.__spec__.
+        loader_state = {}
+        loader_state["__dict__"] = module.__dict__.copy()
+        loader_state["__class__"] = module.__class__
+        module.__spec__.loader_state = loader_state
+        module.__class__ = DelayedImportErrorModule
+
+
+class DelayedImportErrorModule(types.ModuleType):
+    """A subclass of ModuleType which trigger an ImportError upon attribute access
+
+    Based on importlib.util._LazyModule
+    """
+
+    def __getattribute__(self, attr):
+        spec = super().__getattribute__("__spec__")
+        raise ModuleNotFoundError(
+            f"Delayed Report: module named '{spec.name}' not found.\n"
+            "Reporting was Lazy -- delayed until module attributes accessed.\n"
+            f"Most likely, {spec.name} is not installed"
+        )
+
+
+def lazy_importorskip(modname, minversion=None, reason=None):
+    module = pytest.importorskip(modname, minversion, reason)
+    if isinstance(module, DelayedImportErrorModule):
+        if reason is None:
+            reason = "Could not import {modname!r}. Lazy import delayed reporting."
+        raise pytest.skip(reason, allow_module_level=True)
+    return module
+
+
+def lazy_import(fullname):
+    """Return a lazily imported proxy for a module or library.
+
+    We often see the following pattern::
+
+      def myfunc():
+          import scipy
+          ....
+
+    This is to prevent a library, in this case `scipy`, from being
+    imported at function definition time, since that can be slow.
+    This function provides a proxy module that, upon access, imports
+    the actual module.
+
+    Parameters
+    ----------
+
+    fullname : str
+        The full name of the package or subpackage to import.  For example::
+          sp = lazy_import('scipy')  # import scipy as sp
+          spla = lazy_import('scipy.linalg')  # import scipy.linalg as spla
+
+    Returns
+    -------
+
+    pm : importlib.util._LazyModule
+        Proxy module. Can be used like any regularly imported module.
+    """
+    try:
+        return sys.modules[fullname]
+    except:
+        pass
+
+    # Not previously loaded -- look it up
+    spec = importlib.util.find_spec(fullname)
+
+    if spec is not None:
+        module = importlib.util.module_from_spec(spec)
+
+        # Make module with proper locking and get it inserted into sys.modules.
+        loader = importlib.util.LazyLoader(spec.loader)
+        sys.modules[fullname] = module
+        loader.exec_module(module)
+        return module
+
+    # package not found - construct delayed error module
+    spec = importlib.util.spec_from_loader(fullname, loader=None)
+    module = importlib.util.module_from_spec(spec)
+    tmp_loader = importlib.machinery.SourceFileLoader(module, path=None)
+    loader = DelayedImportErrorLoader(tmp_loader)
+
+    sys.modules[fullname] = module
```

Comment:
Do we want to insert modules not found into `sys.modules`?

Reply:
This is the tricky part -- I couldn't figure out another way to make future imports work. But I thought of something yesterday which make removing this more feasible:

Perhaps we should make delayed_not_found_errors only work for the current module's scope. That is, they return a module object (lazily loaded) but don't change `sys.modules`.  Then `np` is available in the module as a trigger for a delayed error report, but it doesn't cause delayed error reports for other modules (or other libraries who don't know anything about networkx's lazy load setup -- which I think is a much bigger problem).  

It does mean that we have to lazy load, check for `spec`, catch `spec is None`, and create this DelayedModuleNotFound object inside each module where we use `numpy` when `numpy` is not installed. But I'm guessing that is not going to be a performance problem.

Not adding to `sys.modules` might also remove any need to overcome `importorskip`.
Yes!  I think this will be a cleaner approach -- if I'm not missing anything.  I'll try it. :}